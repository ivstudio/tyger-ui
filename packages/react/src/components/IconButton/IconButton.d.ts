export type IconsButtonsSizes = 'sm' | 'md' | 'lg';

export interface IconRootProps {
    size?: IconsButtonsSizes;
    color?: string;
    background?: string;
    filled?: boolean;
}

export interface IconButtonProps extends IconRootProps {
    children: React.ReactNode;
    onClick?: () => void;
    disabled?: boolean;
    className?: string;
}
