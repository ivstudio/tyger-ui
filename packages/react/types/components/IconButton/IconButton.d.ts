/// <reference types="react" />
export declare type TSize = 'sm' | 'md' | 'lg';
interface IIconRoot {
    size?: TSize;
    color?: string;
    background?: string;
    filled?: boolean;
}
interface IIconButton extends IIconRoot {
    children: React.ReactNode;
    onClick?: () => void;
    disabled?: boolean;
    className?: string;
}
export declare const IconButtonRoot: import("styled-components").StyledComponent<"button", import("styled-components").DefaultTheme, IIconRoot, never>;
declare const IconButton: ({ children, onClick, size, disabled, background, color, filled, className, ...baseProps }: IIconButton) => JSX.Element;
export default IconButton;
