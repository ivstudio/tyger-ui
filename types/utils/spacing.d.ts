export declare const spacing: {
    readonly '0': "0px";
    readonly '2': "2px";
    readonly '4': "4px";
    readonly '8': "8px";
    readonly '12': "12px";
    readonly '16': "16px";
    readonly '24': "24px";
    readonly '32': "32px";
    readonly '40': "40px";
    readonly '48': "48px";
    readonly '64': "64px";
    readonly '80': "80px";
    readonly '96': "96px";
    readonly '160': "160px";
};
export declare type TSpacing = keyof typeof spacing;
