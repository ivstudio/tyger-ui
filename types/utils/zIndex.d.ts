export declare const zIndex: {
    readonly mobileStepper: 1000;
    readonly appBar: 1100;
    readonly drawer: 1200;
    readonly modal: 1300;
    readonly snackbar: 1400;
    readonly tooltip: 1500;
};
export declare type TZIndex = keyof typeof zIndex;
