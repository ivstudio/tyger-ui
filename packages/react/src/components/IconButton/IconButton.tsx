'use client';
import styled, { useTheme } from 'styled-components';

import {
    backgroundTransition,
    buttonBase,
    flexCenterInline,
} from '../../styles';
import {
    IconButtonProps,
    IconRootProps,
    IconsButtonsSizes,
} from './IconButton.d';

const getSize = (size: IconsButtonsSizes = 'sm') => {
    const buttonSizes = {
        sm: 4,
        md: 6,
        lg: 8,
    };
    const iconSizes = {
        sm: 24,
        md: 32,
        lg: 40,
    };
    return {
        padding: buttonSizes[size],
        fontSize: iconSizes[size],
    };
};

export const IconButtonRoot = styled.button<IconRootProps>`
    ${({ size }) => getSize(size)};
    color: ${({ color }) => color};
    background-color: ${({ background, filled }) => {
        return filled ? background : 'transparent';
    }};
    border-radius: 50%;
    overflow: hidden;
    ${buttonBase};
    ${backgroundTransition};
    ${flexCenterInline};
    -webkit-tap-highlight-color: transparent;

    &:hover {
        background-color: ${({ background }) => background};
    }

    &:disabled {
        color: ${({ theme: { iconButton } }) => iconButton.disabled.color};
    }
`;

const IconButton = ({
    children,
    onClick,
    size = 'sm',
    disabled = false,
    background,
    color,
    filled = false,
    className,
    ...baseProps
}: IconButtonProps) => {
    const { iconButton } = useTheme();

    return (
        <IconButtonRoot
            {...baseProps}
            className={className}
            onClick={onClick}
            size={size}
            disabled={disabled}
            background={background || iconButton.hover.backgroundColor}
            color={color || iconButton.color}
            filled={filled}
        >
            {children}
        </IconButtonRoot>
    );
};

export default IconButton;
