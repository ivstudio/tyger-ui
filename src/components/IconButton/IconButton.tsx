import styled, { useTheme } from 'styled-components';

import {
    backgroundTransition,
    buttonBase,
    flexCenterInline,
} from '../../styles';

export type TSize = 'sm' | 'md' | 'lg';

interface IIconRoot {
    size?: TSize;
    color?: string;
    background?: string;
    filled?: boolean;
}
interface IIconButton extends IIconRoot {
    children: React.ReactNode | React.ReactNode[];
    onClick?: () => void;
    disabled?: boolean;
    className?: string;
}

const getSize = (size: TSize = 'sm') => {
    const buttonSizes = {
        sm: 4,
        md: 6,
        lg: 8,
    };
    const iconSizes = {
        sm: 24,
        md: 32,
        lg: 40,
    };
    return {
        padding: buttonSizes[size],
        fontSize: iconSizes[size],
    };
};

export const IconButtonRoot = styled.button<IIconRoot>`
    ${({ size }) => getSize(size)};
    color: ${({ color }) => color};
    background-color: ${({ background, filled }) => {
        return filled ? background : 'transparent';
    }};
    border-radius: 50%;
    overflow: hidden;
    ${buttonBase};
    ${backgroundTransition};
    ${flexCenterInline};
    -webkit-tap-highlight-color: transparent;

    &:hover {
        background-color: ${({ background }) => background};
    }
`;

const IconButton = ({
    children,
    onClick,
    size = 'sm',
    disabled = false,
    background,
    color,
    filled = false,
    className,
    ...baseProps
}: IIconButton) => {
    const { palette } = useTheme();

    return (
        <IconButtonRoot
            {...baseProps}
            className={className}
            onClick={onClick}
            size={size}
            disabled={disabled}
            background={background || palette.ui.primary}
            color={color || palette.text.secondary}
            filled={filled}
        >
            {children}
        </IconButtonRoot>
    );
};

export default IconButton;
