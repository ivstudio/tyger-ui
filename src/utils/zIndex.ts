export const zindex = {
    mobileStepper: 1000,
    appBar: 1100,
    drawer: 1200,
    modal: 1300,
    snackbar: 1400,
    tooltip: 1500,
};
export type TZIndex = keyof typeof zindex;
