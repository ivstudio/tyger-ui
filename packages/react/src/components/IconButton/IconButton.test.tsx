import React from 'react';
import IconButton from './IconButton';
import { MdOutlineMenu as MenuIcon } from 'react-icons/md';
import { render, fireEvent, screen } from '../../../test/utils/mockProvider';

const clickHandler = jest.fn();

describe('IconButton', () => {
    it('renders IconButton correctly', () => {
        const { container } = render(
            <IconButton filled size="sm" onClick={clickHandler}>
                <MenuIcon />
            </IconButton>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    it('handles click event', () => {
        render(
            <IconButton filled size="sm" onClick={clickHandler}>
                <MenuIcon />
            </IconButton>
        );

        fireEvent.click(screen.getByRole('button'));
        expect(clickHandler).toHaveBeenCalledTimes(1);
    });

    it('will not click when it is disabled', () => {
        render(
            <IconButton filled size="sm" onClick={clickHandler} disabled>
                <MenuIcon />
            </IconButton>
        );
        const buttonElem = screen.getByRole('button');

        expect(buttonElem).toBeDisabled();

        fireEvent.click(buttonElem);
        expect(clickHandler).toHaveBeenCalledTimes(0);
    });
});
