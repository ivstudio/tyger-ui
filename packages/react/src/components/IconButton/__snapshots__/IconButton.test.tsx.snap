// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton renders IconButton correctly 1`] = `
<button
  class="sc-aXZVg doMpLJ"
  color="#fafafa"
>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;
